### ISSUE TITLE
ss-redir gets stuck in getsockopt() on Linux 4.15.6

### ISSUE THREAD

**Issue by PeterCxy** [shadowsocks/shadowsocks-libev#1955]
### What version of shadowsocks-libev are you using?

shadowsocks-libev 3.1.3

### What operating system are you using?

ArchLinux

### What did you do?

Installed Linux 4.15.6 (everything worked fine before this update)

### What did you expect to see?

ss-redir works

### What did you see instead?

ss-redir process gets stuck. Even Ctrl+Z can't kill the process (got stuck in `getsockopt()`)

Deleting calls to `getsockopt()` in `redir.c` and the process won't get stuck, but `ss-redir` depends on these stuff to work :(

This issue was also reported by other users using Linux 4.15.6, mostly in `archlinux-cn`, not sure whether it is a kernel bug, or bug of ArchLinux, or bug of shadowsocks-libev.

**Comment by PeterCxy** [comment]
Here is debug information dumped from ss-redir (provided by Jerry981028)

https://gist.github.com/Jerry981028/6668324201aae189915634b35281b9ca

Most noticably,

```
[  246.751395] Call Trace:
[  246.751403]  ? __schedule+0x24b/0x8c0
[  246.751406]  ? ep_send_events_proc+0x7f/0x1a0
[  246.751408]  schedule+0x32/0x90
[  246.751410]  __lock_sock+0x79/0xc0
[  246.751413]  ? wait_woken+0x80/0x80
[  246.751415]  lock_sock_nested+0x50/0x60
[  246.751419]  getorigdst+0x5a/0x240 [nf_conntrack_ipv4]
[  246.751421]  ? preempt_count_add+0x49/0xa0
[  246.751424]  nf_getsockopt+0x47/0x70
[  246.751426]  ip_getsockopt+0x7f/0xc0
[  246.751429]  SyS_getsockopt+0x76/0xd0
[  246.751431]  do_syscall_64+0x74/0x190
[ 246.751433] entry_SYSCALL_64_after_hwframe+0x3d/0xa2
```

**Comment by chih7** [comment]
Same bug in Linux  4.14.23.

**Comment by PeterCxy** [comment]
This was due to a deadlock introduced by linux 4.15.6. Fixed in 4.15.7. Closing this issue.
